Ncsdk 4093 add support for supl client and update gps sample

evenl opened the pull request:
This will allow users to use the SUPL client library in NCS and will also update the current GPS example to also support using the SUPL client library to fetch AGPS data. Because of license restrictions will the SUPL client library be distributed in a separate zip file, which can be downloaded from the nordicsemic.com web page after a registration process (not available yet). 

The zip file is to be downloaded and then unzipped in the root of the NCS folder alongside the other folders (nrf, zephyr, mcuboot, nrfxlib, etc). This will create a folder called supl, where the .a file + some license text, readme document will reside. The purpose was to try to make it as user friendly as possible, in stead of instructing the user to locate some folder deep inside some repository. Which would have confused west/git as the repo would then contain new un-tracked files.

NordicBuilder commented:
**Some checks failed. Please fix and resubmit.**

## checkpatch issues

```
No additional types will be considered - file 'scripts/checkpatch/typedefsfile': No such file or directory
-:134: ERROR:POINTER_LOCATION: "foo * bar" should be "foo *bar"
#134: FILE: lib/bin/supllib/include/supl_os_client.h:4:
+int nrf_supl_session(nrf_gnss_agps_data_frame_t * agps_request);

-:185: WARNING:LONG_LINE_COMMENT: line over 80 characters
#185: FILE: lib/bin/supllib/include/supl_session.h:44:
+ *    “DOD”), the software is classified as “Commercial Computer Programs” and

-:192: WARNING:LONG_LINE_COMMENT: line over 80 characters
#192: FILE: lib/bin/supllib/include/supl_session.h:51:
+ *    agency other than DOD, the software is classified as “Commercial Computer

-:194: WARNING:LONG_LINE_COMMENT: line over 80 characters
#194: FILE: lib/bin/supllib/include/supl_session.h:53:
+ *    classified as “Commercial Items” and pursuant to FAR Section 12.212, the

-:230: WARNING:LONG_LINE_COMMENT: line over 80 characters
#230: FILE: lib/bin/supllib/include/supl_session.h:89:
+ * THIRD PARTY LICENSORS’ LIABILITY SHALL IN NO CIRCUMSTANCE EXCEED AN AGGREGATE

-:317: WARNING:NEW_TYPEDEFS: do not add new typedefs
#317: FILE: lib/bin/supllib/include/supl_session.h:176:
+typedef struct lte_params {

-:327: WARNING:NEW_TYPEDEFS: do not add new typedefs
#327: FILE: lib/bin/supllib/include/supl_session.h:186:
+typedef struct supl_session_ctx {

-:335: WARNING:LONG_LINE_COMMENT: line over 80 characters
#335: FILE: lib/bin/supllib/include/supl_session.h:194:
+	void *user_data;                          /* OPTIONAL argument for callbacks */

-:342: WARNING:LONG_LINE_COMMENT: line over 80 characters
#342: FILE: lib/bin/supllib/include/supl_session.h:201:
+ * @param[in]   nbytes      Number of bytes to read. Shall not be larger than the size of p_buff.

-:411: WARNING:NEW_TYPEDEFS: do not add new typedefs
#411: FILE: lib/bin/supllib/include/supl_session.h:270:
+typedef struct supl_api {

-:419: WARNING:NEW_TYPEDEFS: do not add new typedefs
#419: FILE: lib/bin/supllib/include/supl_session.h:278:
+typedef struct supl_buffers {

-:632: WARNING:NAKED_SSCANF: unchecked sscanf return value
#632: FILE: lib/bin/supllib/os/lte_params.c:47:
+	sscanf(&p_buf[buf_len - 1], "%d", &status);

-:632: WARNING:SSCANF_TO_KSTRTO: Prefer kstrto<type> to single variable sscanf
#632: FILE: lib/bin/supllib/os/lte_params.c:47:
+	sscanf(&p_buf[buf_len - 1], "%d", &status);

-:667: WARNING:CONSTANT_COMPARISON: Comparisons should place the constant on the right side of the test
#667: FILE: lib/bin/supllib/os/lte_params.c:82:
+	if (0 != char2hex(p_buf[1], &c)) {

-:673: WARNING:CONSTANT_COMPARISON: Comparisons should place the constant on the right side of the test
#673: FILE: lib/bin/supllib/os/lte_params.c:88:
+	if (0 != char2hex(p_buf[2], &c)) {

-:679: WARNING:CONSTANT_COMPARISON: Comparisons should place the constant on the right side of the test
#679: FILE: lib/bin/supllib/os/lte_params.c:94:
+	if (0 != char2hex(p_buf[3], &c)) {

-:710: WARNING:CONSTANT_COMPARISON: Comparisons should place the constant on the right side of the test
#710: FILE: lib/bin/supllib/os/lte_params.c:125:
+	if (0 != char2hex(p_buf[4], &c)) {

-:716: WARNING:CONSTANT_COMPARISON: Comparisons should place the constant on the right side of the test
#716: FILE: lib/bin/supllib/os/lte_params.c:131:
+	if (0 != char2hex(p_buf[5], &c)) {

-:723: WARNING:CONSTANT_COMPARISON: Comparisons should place the constant on the right side of the test
#723: FILE: lib/bin/supllib/os/lte_params.c:138:
+	if (MIN_XMONITOR_MCC_MNC_LEN == buf_len) {

-:727: WARNING:CONSTANT_COMPARISON: Comparisons should place the constant on the right side of the test
#727: FILE: lib/bin/supllib/os/lte_params.c:142:
+		if (0 != char2hex(p_buf[6], &c)) {

-:772: WARNING:LONG_LINE_COMMENT: line over 80 characters
#772: FILE: lib/bin/supllib/os/lte_params.c:187:
+ *                          a 4-byte E-UTRAN cell ID in hexadecimal format with quotes.

-:831: WARNING:LONG_LINE: line over 80 characters
#831: FILE: lib/bin/supllib/os/lte_params.c:246:
+		return parse_lte_status(&p_session_ctx->lte_params, p_buf, buf_len);

-:832: WARNING:UNNECESSARY_BREAK: break is not useful after a goto or return
#832: FILE: lib/bin/supllib/os/lte_params.c:247:
+		return parse_lte_status(&p_session_ctx->lte_params, p_buf, buf_len);
+		break;

-:834: WARNING:LONG_LINE: line over 80 characters
#834: FILE: lib/bin/supllib/os/lte_params.c:249:
+		if (0 != parse_lte_mcc(&p_session_ctx->lte_params, p_buf, buf_len)) {

-:834: WARNING:CONSTANT_COMPARISON: Comparisons should place the constant on the right side of the test
#834: FILE: lib/bin/supllib/os/lte_params.c:249:
+		if (0 != parse_lte_mcc(&p_session_ctx->lte_params, p_buf, buf_len)) {

-:837: WARNING:LONG_LINE: line over 80 characters
#837: FILE: lib/bin/supllib/os/lte_params.c:252:
+		if (0 != parse_lte_mnc(&p_session_ctx->lte_params, p_buf, buf_len)) {

-:837: WARNING:CONSTANT_COMPARISON: Comparisons should place the constant on the right side of the test
#837: FILE: lib/bin/supllib/os/lte_params.c:252:
+		if (0 != parse_lte_mnc(&p_session_ctx->lte_params, p_buf, buf_len)) {

-:842: WARNING:LONG_LINE: line over 80 characters
#842: FILE: lib/bin/supllib/os/lte_params.c:257:
+		return parse_lte_tac(&p_session_ctx->lte_params, p_buf, buf_len);

-:843: WARNING:UNNECESSARY_BREAK: break is not useful after a goto or return
#843: FILE: lib/bin/supllib/os/lte_params.c:258:
+		return parse_lte_tac(&p_session_ctx->lte_params, p_buf, buf_len);
+		break;

-:845: WARNING:LONG_LINE: line over 80 characters
#845: FILE: lib/bin/supllib/os/lte_params.c:260:
+		return parse_lte_cid(&p_session_ctx->lte_params, p_buf, buf_len);

-:846: WARNING:UNNECESSARY_BREAK: break is not useful after a goto or return
#846: FILE: lib/bin/supllib/os/lte_params.c:261:
+		return parse_lte_cid(&p_session_ctx->lte_params, p_buf, buf_len);
+		break;

-:848: WARNING:LONG_LINE: line over 80 characters
#848: FILE: lib/bin/supllib/os/lte_params.c:263:
+		return parse_lte_phys_cid(&p_session_ctx->lte_params, p_buf, buf_len);

-:849: WARNING:UNNECESSARY_BREAK: break is not useful after a goto or return
#849: FILE: lib/bin/supllib/os/lte_params.c:264:
+		return parse_lte_phys_cid(&p_session_ctx->lte_params, p_buf, buf_len);
+		break;

-:916: WARNING:CONSTANT_COMPARISON: Comparisons should place the constant on the right side of the test
#916: FILE: lib/bin/supllib/os/supl_os_client.c:55:
+		if (NULL == col_end) {

-:923: WARNING:CONSTANT_COMPARISON: Comparisons should place the constant on the right side of the test
#923: FILE: lib/bin/supllib/os/supl_os_client.c:62:
+		if (0 != handler(p_session_ctx, col_number, idx, col_len)) {

-:927: WARNING:CONSTANT_COMPARISON: Comparisons should place the constant on the right side of the test
#927: FILE: lib/bin/supllib/os/supl_os_client.c:66:
+		if (NULL == col_end) {

-:947: WARNING:CONSTANT_COMPARISON: Comparisons should place the constant on the right side of the test
#947: FILE: lib/bin/supllib/os/supl_os_client.c:86:
+	if (SUPL_IMSI_LENGTH != hexstr2hex(response,

-:948: WARNING:LONG_LINE: line over 80 characters
#948: FILE: lib/bin/supllib/os/supl_os_client.c:87:
+					   get_line_len(response, sizeof(response)),

-:964: WARNING:LONG_LINE: line over 80 characters
#964: FILE: lib/bin/supllib/os/supl_os_client.c:103:
+	if (parse_at_resp(&supl_client_ctx, response, sizeof(response), handler_at_xmonitor)) {

-:971: ERROR:POINTER_LOCATION: "foo * bar" should be "foo *bar"
#971: FILE: lib/bin/supllib/os/supl_os_client.c:110:
+int nrf_supl_session(nrf_gnss_agps_data_frame_t * agps_request)

-:984: ERROR:OPEN_BRACE: that open brace { should be on the previous line
#984: FILE: lib/bin/supllib/os/supl_os_client.c:123:
+	if (supl_init(api, &bufs) != 0)
+	{

-:1027: WARNING:LINE_SPACING: Missing a blank line after declarations
#1027: FILE: lib/bin/supllib/os/utils.c:30:
+		u8_t c;
+		if (0 != char2hex(in[i], &c)) {

-:1027: WARNING:CONSTANT_COMPARISON: Comparisons should place the constant on the right side of the test
#1027: FILE: lib/bin/supllib/os/utils.c:30:
+		if (0 != char2hex(in[i], &c)) {

-:1238: ERROR:OPEN_BRACE: that open brace { should be on the previous line
#1238: FILE: samples/nrf9160/gps/src/main.c:87:
+	if (activate)
+	{

-:1240: ERROR:OPEN_BRACE: that open brace { should be on the previous line
#1240: FILE: samples/nrf9160/gps/src/main.c:89:
+		if (at_cmd_write(AT_ACTIVATE_LTE, NULL, 0, NULL) != 0)
+		{

-:1248: ERROR:OPEN_BRACE: that open brace { should be on the previous line
#1248: FILE: samples/nrf9160/gps/src/main.c:95:
+		if (at_cmd_write("AT+CEREG=2", NULL, 0, NULL) != 0)
+		{

-:1261: ERROR:OPEN_BRACE: that open brace { should be on the previous line
#1261: FILE: samples/nrf9160/gps/src/main.c:103:
+		if (at_cmd_write("AT+CEREG=0", NULL, 0, NULL) != 0)
+		{

-:1266: ERROR:OPEN_BRACE: that open brace { should be on the previous line
#1266: FILE: samples/nrf9160/gps/src/main.c:108:
+		if (at_cmd_write(AT_DEACTIVATE_LTE, NULL, 0, NULL) != 0)
+		{

-:1337: WARNING:LONG_LINE: line over 80 characters
#1337: FILE: samples/nrf9160/gps/src/main.c:252:
+	retval = nrf_recv(gnss_fd, gps_data, sizeof(nrf_gnss_data_frame_t), NRF_MSG_DONTWAIT);

-:1349: WARNING:LONG_LINE: line over 80 characters
#1349: FILE: samples/nrf9160/gps/src/main.c:287:
+			printk("New AGPS data requested, contacting SUPL server, flags %d\n", gps_data->agps.data_flags);

-:1385: WARNING:LONG_LINE: line over 80 characters
#1385: FILE: samples/nrf9160/gps/src/main.c:324:
+		printk("Failed to send AGNSS data, type: %d (err: %d)\n", type, errno);

-:1387: WARNING:UNNECESSARY_ELSE: else is not generally useful after a break or return
#1387: FILE: samples/nrf9160/gps/src/main.c:326:
+		return -1;
+	} else {

-:1388: WARNING:LONG_LINE: line over 80 characters
#1388: FILE: samples/nrf9160/gps/src/main.c:327:
+		printk("Injected AGPS data, flags: %d, size: %d\n", type, agps_size);

-:1418: WARNING:LINE_SPACING: Missing a blank line after declarations
#1418: FILE: samples/nrf9160/gps/src/main.c:357:
+	int rc = nrf_supl_init(&supl_api);
+	if (rc != 0)

-:1418: ERROR:OPEN_BRACE: that open brace { should be on the previous line
#1418: FILE: samples/nrf9160/gps/src/main.c:357:
+	if (rc != 0)
+	{

-:1458: WARNING:BLOCK_COMMENT_STYLE: Block comments use * on subsequent lines
#1458: FILE: samples/nrf9160/gps/src/supl_support.c:26:
+		/* Either a valid,
+		   NULL-terminated access point name or NULL. */

-:1458: WARNING:BLOCK_COMMENT_STYLE: Block comments use a trailing */ on a separate line
#1458: FILE: samples/nrf9160/gps/src/supl_support.c:26:
+		   NULL-terminated access point name or NULL. */

-:1459: ERROR:C99_COMMENTS: do not use C99 // comments
#1459: FILE: samples/nrf9160/gps/src/supl_support.c:27:
+		.ai_canonname = NULL// (char *)cfg->apn

-:1504: WARNING:LINE_SPACING: Missing a blank line after declarations
#1504: FILE: samples/nrf9160/gps/src/supl_support.c:72:
+			char ip[255] = { 0 };
+			nrf_inet_ntop(NRF_AF_INET,

-:1505: WARNING:LONG_LINE: line over 80 characters
#1505: FILE: samples/nrf9160/gps/src/supl_support.c:73:
+				      (void *)&((struct nrf_sockaddr_in *)sa)->sin_addr,

-:1515: ERROR:C99_COMMENTS: do not use C99 // comments
#1515: FILE: samples/nrf9160/gps/src/supl_support.c:83:
+		//log_addrinfo(addr);

-:1559: WARNING:LINE_SPACING: Missing a blank line after declarations
#1559: FILE: samples/nrf9160/gps/src/supl_support.c:127:
+	int ret = vsnprintk(buffer, sizeof(buffer), fmt, args);
+	va_end(args);

-:1562: WARNING:EMBEDDED_FUNCTION_NAME: Prefer using '"%s...", __func__' to using 'logger', this function's name, in a string
#1562: FILE: samples/nrf9160/gps/src/supl_support.c:130:
+		printk("logger: encoding error\n");

-:1565: WARNING:EMBEDDED_FUNCTION_NAME: Prefer using '"%s...", __func__' to using 'logger', this function's name, in a string
#1565: FILE: samples/nrf9160/gps/src/supl_support.c:133:
+		printk("logger: too long message, it will be cut short\n");

-:1568: ERROR:C99_COMMENTS: do not use C99 // comments
#1568: FILE: samples/nrf9160/gps/src/supl_support.c:136:
+	//printk("%s", log_strdup(buffer));

-:1578: WARNING:CONSTANT_COMPARISON: Comparisons should place the constant on the right side of the test
#1578: FILE: samples/nrf9160/gps/src/supl_support.c:146:
+	if (rc < 0 && (ETIMEDOUT == errno)) {

-:1605: ERROR:TRAILING_WHITESPACE: trailing whitespace
#1605: FILE: samples/nrf9160/gps/src/supl_support.h:16:
+ * @brief This function will close the socket previously by the $

-:1612: ERROR:TRAILING_WHITESPACE: trailing whitespace
#1612: FILE: samples/nrf9160/gps/src/supl_support.h:23:
+ * $

- total: 14 errors, 54 warnings, 1546 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

NOTE: Whitespace errors detected.
      You may wish to use scripts/cleanpatch or scripts/cleanfile

Your patch has style problems, please review.

NOTE: Ignored message types: AVOID_EXTERNS BRACES CONFIG_EXPERIMENTAL CONST_STRUCT DATE_TIME FILE_PATH_CHANGES MINMAX MULTISTATEMENT_MACRO_USE_DO_WHILE NETWORKING_BLOCK_COMMENT_STYLE PRINTK_WITHOUT_KERN_LEVEL SPLIT_STRING VOLATILE

NOTE: If any of the errors are false positives, please report
      them to the maintainers.
```

## Identity/Emails issues

3ee944398f5edfce75e4d486315b0f432b712a52: author email (Even Falch-Larsen <even.falch-larsen@nordicsemi.no>) needs to match one of the signed-off-by entries.



*Tip: The bot edits this comment instead of posting a new one, so you can check the comment's history to see earlier messages.*

rlubos commented:
@carlescufi Before we proceed with the actual review,

> Because of license restrictions will the SUPL client library be distributed in a separate zip file, which can be downloaded from the nordicsemic.com web page after a registration process (not available yet).

This is something new within the NCS project, are you ok with this? Do we have an approval to proceed with this PR, or are there any other ideas on how to distribute the sample?

jhn-nordic commented:
> @carlescufi Before we proceed with the actual review,
> 
> > Because of license restrictions will the SUPL client library be distributed in a separate zip file, which can be downloaded from the nordicsemic.com web page after a registration process (not available yet).
> 
> This is something new within the NCS project, are you ok with this? Do we have an approval to proceed with this PR, or are there any other ideas on how to distribute the sample?

The Bosch BSEC library we use on Thingy:91 is quite similar. That needs to be put into nrf/ext/

metc commented on samples/nrf9160/gps/src/main.c:
```diff
@@ -37,44 +53,59 @@ static bool           update_terminal;
 
-		if (bytes_sent < 0) {
-			close(at_sock);
+static void wait_for_lte(char *response)
```
I think using the link controller module could help for both functions `wait_for_lte` and `activate_lte`.

evenl replied:
I had a chat with Jan Tore, one of the authors of the link controller, but it doesn't seem like we are able to get it to work with this use-case. So I stick with my current implementation and then we can update it later when we have more time for debugging.

rlubos commented on lib/bin/supllib/os/lte_params.c:
```diff
@@ -0,0 +1,284 @@
+/**
+ * @brief Parse and fill XMONITOR physical CID to client context lte_params
+ *
+ * @param[in,out] p_session_ctx pointer to client context to fill
```
It's `p_lte`.

evenl replied:
fixed

rlubos commented on lib/bin/supllib/os/supl_os_client.c:
```diff
@@ -0,0 +1,132 @@
+#include <logging/log.h>
```
Missing license header.

evenl replied:
ok

rlubos commented on samples/nrf9160/gps/prj.conf:
```diff
@@ -5,14 +5,38 @@
+CONFIG_NET_SOCKETS_POSIX_NAMES=y
+CONFIG_NET_SOCKETS_SOCKOPT_TLS=y
+
+# Disable native network stack to save some memory
```
Use `CONFIG_NET_NATIVE=n` instead.

evenl replied:
ok, I assume that is for all of the CONFIG_NET_* lines below line 25.

rlubos commented on samples/nrf9160/gps/prj.conf:
```diff
@@ -5,14 +5,38 @@
+CONFIG_NET_TCP=n
+
+# Main thread
+CONFIG_MAIN_THREAD_PRIORITY=0
```
It's the default value, no need to redefine it.

evenl replied:
ok

rlubos commented on samples/nrf9160/gps/prj.conf:
```diff
@@ -5,14 +5,38 @@
 #
```
I believe that the sample documentation should be updated as well, to indicate the SUPL library support.

evenl replied:
I think we need to write the documentation in the first place.

rlubos commented on lib/bin/supllib/readme.txt:
```diff
@@ -0,0 +1,18 @@
+This is the SUPL Client library component that adds SUPL support to NCS. The
```
Shouldn't this be in `.rst` format, not `.txt`?

evenl replied:
Yes

rlubos commented on samples/nrf9160/gps/src/main.c:
```diff
@@ -37,44 +53,59 @@ static bool           update_terminal;
-		do {
-			bytes_received = recv(at_sock, buf, 2, 0);
-		} while (bytes_received == 0);
+		at_cmd_set_notification_handler(wait_for_lte);
```
You should use the `at_notif` API to register a callback.

evenl replied:
Ok, was not aware that this was done.

evenl replied:
Code updated

carlescufi commented on lib/bin/supllib/include/supl_session.h:
```diff
@@ -0,0 +1,317 @@
+ * Copyright (c) 2019 - 2019, Nordic Semiconductor ASA
+ * All rights reserved.
+ *
+ * 1. By downloading this software, you acknowledge that all right, title and
```
Is it really necessary to have this license here instead of the standard Nordic 5-clause? CC @bjorn-tore-taraldsen

bjorn-tore-taraldsen replied:
The library (.a) has to be distributed with this license and this is the .h file that get's delivered with this library (.a). It does not make sense to have different license on these two files.

rlubos commented on samples/nrf9160/gps/src/main.c:
```diff
@@ -244,6 +280,25 @@ int process_gps_data(nrf_gnss_data_frame_t *gps_data)
+			printk("New AGPS data requested, contacting SUPL server, flags %d\n",
+			       gps_data->agps.data_flags);
+			activate_lte(true);
+			printf("Established LTE link\n");
```
`printk` here and below.

evenl replied:
ok

carlescufi commented on lib/bin/supllib/license.txt:
```diff
@@ -0,0 +1,70 @@
+Copyright (c) 2019 - 2019, Nordic Semiconductor ASA
+All rights reserved.
+
+1.  By downloading this software, you acknowledge that all right, title and interest in and to the
```
Again, do we need this license here? The file itself (the library) is not here, so I am not sure it makes sense at all. The license should be displayed and shown when downloading the library from Nordic's website, but not included here? cc @bjorn-tore-taraldsen

evenl replied:
I agree that including the license here is a bit strange as there is nothing to license here.

bjorn-tore-taraldsen replied:
We could be okay with only having the license in the .h file, but we need to visualize for the user that this library comes with a different license.

carlescufi commented on lib/bin/supllib/CMakeLists.txt:
```diff
@@ -0,0 +1,38 @@
+
+set(lib_path lib/${GCC_M_CPU}/${float_dir})
+
+set(SUPL_CLIENT_LIB_PATH ${ZEPHYR_BASE}/../supl/${lib_path})
```
Instead of forcing it to a path, should we use an environment variable instead? CC @SebastianBoe and @bjorn-tore-taraldsen

carlescufi replied:
If we have to choose a path,  then I would probably suggest instead what @jhn-nordic proposes: `nrf/ext`

evenl replied:
I think forcing a path is the simples solution for the user to use, and for us to explain in the documentation. But it also give user the fewest possibilities to configure their repo by them self. So the question here is probably what are the chances that the user want this .a file to be located somewhere else.

evenl replied:
Ok, I thought  putting it into nrf/ext would confuse west, but that is of course for versioned files. Yes, I can move it in to that folder. So I just need conclution on foced vs environment variable.

bjorn-tore-taraldsen replied:
Wouldn't placing it inside NCS folder structure, make it easier for the user to, by accident, publish the library as part of their git repo, and by that make it public available? By doing this the user are breaking the license agreement. 
My thinking is having it located "on the side of" the repo structure could prevent this from happening.

SebastianBoe replied:
libsupl.a should be in the variable as well, and it should be overridable through an env var.

carlescufi replied:
Yes @bjorn-tore-taraldsen but see my comment [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1792#issuecomment-579274043). Adding it to the `.gitignore` would make it relatively hard for the users to do this by mistake.

evenl replied:
I have now moved the location of the library to ext/supl and added a .gitignore file to ignore the supl folder. I have also removed the supl_session.h file from the nrf integration code in this PR and it will only be distributed in the .zip file that the user download.

Because I moved the expected location of the supl library into the nrf/ext/ folder, I changed the CMakeLists.txt file in that folder to look for the supl folder if the CONFIG_SUPL_CLIENT_LIB is set. And then I added a CMakeLists.txt file to the .zip file that we are going to distribute. This simplifies CMakeLists.txt files and makes them more maintainable.

rlubos commented on lib/bin/supllib/include/supl_session.h:
```diff
@@ -0,0 +1,317 @@
+/*
+ * Copyright (c) 2019 - 2019, Nordic Semiconductor ASA
```
This header file defines the library API, IMO it should be distributed with the library.

evenl replied:
It's distributed with the library, I don't remember why we ended up with a copy here. I can remove it.

evenl replied:
I removed the supl_session.h from the nrf part, this file will only be distributed with the .zip package.

rlubos commented on samples/nrf9160/gps/src/supl_support.c:
```diff
@@ -0,0 +1,154 @@
+	struct nrf_addrinfo *info;
+
+	proto = NRF_IPPROTO_TCP;
+	port = nrf_htons(7276);
```
The port number could have it's #define, just below `SUPL_SERVER`

evenl replied:
yes

rlubos commented on samples/nrf9160/gps/src/supl_support.c:
```diff
@@ -0,0 +1,154 @@
+		return -1;
+	}
+
+	supl_fd = nrf_socket(NRF_AF_INET, NRF_SOCK_STREAM, proto);
```
We do plain TCP/IP here which is fully offloaded, I suggest to use the posix API then.

evenl replied:
TCP code updated to use offloaded socket API

rlubos commented on samples/nrf9160/gps/src/supl_support.h:
```diff
@@ -0,0 +1,69 @@
+ * @retval 0  If opening socket was successful.
+ *         -1 If opening socket failed
+ */
+int     open_supl_socket(void);
```
Excess spaces here and in the declarations below.

evenl replied:
hmmm, strange, fixed now.

rlubos commented on samples/nrf9160/gps/src/supl_support.h:
```diff
@@ -0,0 +1,69 @@
+ *         and a reference is therefor not passed to the function.
+ *
+ * @retval 0  If opening socket was successful.
+ *         -1 If opening socket failed
```
This needs a spearate `@retval` entry. Here and in the declarations below.

evenl replied:
ok

rlubos commented on lib/bin/supllib/include/supl_os_client.h:
```diff
@@ -0,0 +1,25 @@
+ * @return 0  SUPL session was successful.
+ *         <0 SUPL session failed.
+ */
+int nrf_supl_session(nrf_gnss_agps_data_frame_t *agps_request);
```
Couldn't this be simply `supl_session` and `supl_init`? We tend to avoid `nrf_` prefixing in here.

evenl replied:
Yes, I agree, but the library functions are called that so this would lead to name crash. I can of course rename the functions in the library to something else to free up these names. You thoughts?

rlubos replied:
The header file is `supl_os_client.h` so perhaps `supl_client` would be a good prefix for these functions?

evenl replied:
The library is now exposing supl_client_init, supl_client_session and supl_client_abort. So renamed the functions without the nrf_ prefix.

rlubos commented on samples/nrf9160/gps/src/supl_support.h:
```diff
@@ -0,0 +1,69 @@
+
```
Missing license header and include guard.

evenl replied:
Ok, added the extern c lines too, it's meant as in internal API so probably not necessary. But aligns with the rest

rlubos commented on samples/nrf9160/gps/src/supl_support.c:
```diff
@@ -0,0 +1,154 @@
+#include <zephyr.h>
```
Missing license header.

evenl replied:
ok

rlubos commented on samples/nrf9160/gps/src/supl_support.c:
```diff
@@ -0,0 +1,154 @@
+	err = nrf_getaddrinfo(SUPL_SERVER, NULL, &hints, &info);
+	if (err) {
+		printk("Failed to resolve hostname %s on IPv4, errno: %d)\n",
+			log_strdup(SUPL_SERVER),
```
You should not do this with `printk`.

evenl replied:
ok, removed all log_strdup calls.

rlubos commented on samples/nrf9160/gps/src/supl_support.c:
```diff
@@ -0,0 +1,154 @@
+				      ip,
+				      255);
+			printf("ip %s (%x) port %d\n",
+				log_strdup(ip),
```
You sholud not do this with `printk`

evenl replied:
ok

rlubos replied:
Still not fixed, remove `log_strdup` here and grep for `printf` -> `printk`

evenl replied:
fixed now

rlubos commented on samples/nrf9160/gps/src/supl_support.c:
```diff
@@ -0,0 +1,154 @@
+		       "it will be cut short\n", __func__);
+	}
+
+	printk("%s", log_strdup(buffer));
```
You sholud not do this with `printk`

evenl replied:
ok

evenl commented on lib/bin/supllib/CMakeLists.txt:
```diff
@@ -0,0 +1,38 @@
+
+if(NOT EXISTS ${SUPL_CLIENT_LIB_PATH})
+  message(WARNING "Cannot locate SUPL Client library, should be located in your NCS root/supl. "
+                  "Information about how to download the library can be found at www.nordicsemi.no")
```
My bad

SebastianBoe replied:
I wasn't able to find this, when will it be available?

Also, how many binary variants will it have?

evenl replied:
I don't know the exact time it will be available, but only one binary variant, hard-float for nrf9160.
I can send you a copy of the .a file if you want it for testing purposes?

evenl commented on lib/bin/supllib/Kconfig:
```diff
@@ -0,0 +1,12 @@
+
+menuconfig SUPL_CLIENT_LIB
+	bool "SUPL Client library"
+	default n
```
ok

evenl commented on lib/bin/supllib/include/utils.h:
```diff
@@ -0,0 +1,42 @@
+ * @return  hex output length on success
+ *         -1 on error
+ */
+int hexstr2hex(const char *restrict in,
```
This code I copied from the orginal code
https://www.geeksforgeeks.org/restrict-keyword-c/

Seems to be an optimization hint to the compiler, I can remove it if you want.

SebastianBoe replied:
Could this cause issues for apps that use C++?

rlubos replied:
To be honest, I have learned about this keyword just today so I can't tell from my own experience, but indeed, this bullet point is daunting:
```
* restrict is not supported by C++. It is a C only keyword.
```
Perhaps it would be safer to give up on it if it's really just a copy-paste from some other code.

evenl replied:
This code was written by the thirdpary consultant company that have done GPS work for us. So by copying this code I mean copying form the internal codebase that they have worked on.

SebastianBoe replied:
Understood, but assuming that app C++ sources could directly or indirectly include utils.h, then this could cause issues.

SebastianBoe replied:
Perhaps you could check that we don't accidentally leak utils.h to the app?

evenl replied:
We do not need this keyword, it is strictly an optimization hint to the compiler. This seems like a simpler solution to avoid possible issues.

SebastianBoe replied:
Great, let's remove it then.

evenl replied:
gone

evenl commented on lib/bin/supllib/include/lte_params.h:
```diff
@@ -0,0 +1,34 @@
+ *
+ * @param[in,out]   p_params  LTE parameters to initialize
+ */
+void lte_params_init(struct lte_params *p_params);
```
ok

SebastianBoe commented on lib/bin/supllib/CMakeLists.txt:
```diff
@@ -0,0 +1,38 @@
+  if(CONFIG_FP_HARDABI)
+    set(float_dir hard-float)
+  else()
+    assert(0 "Unreachable code")
```
Is this unreachable code? Could the user not have enabled CONFIG_FP_SOFTABI ?

evenl replied:
we don't support it, the SUPL library is only built for hard-float.
But maybe the message instead should have been "CONFIG_FP_SOFTABI not supported"

SebastianBoe replied:
soft-float is given by not enabling CONFIG_FLOAT. But "Soft float not supported" is not the right error message
because FP_SOFTABI is not soft-float.

The message should be changed to "CONFIG_FP_SOFTABI is not supported".

evenl replied:
Changed the message text, you will not this update as I moved the CMakeLists.txt file into the zipfile that user download.

SebastianBoe commented on lib/bin/supllib/CMakeLists.txt:
```diff
@@ -0,0 +1,38 @@
+                  "Information about how to download the library can be found at www.nordicsemi.no")
+endif()
+
+set(                   SUPL_CLIENT_TARGET ${IMAGE}supl_client)
```
IMAGE can be omitted

evenl replied:
Ok, eradicated

evenl commented on lib/bin/supllib/include/supl_os_client.h:
```diff
@@ -0,0 +1,25 @@
+#include <nrf_socket.h>
```
Ok

SebastianBoe commented on lib/bin/supllib/include/supl_session.h:
```diff
@@ -0,0 +1,317 @@
+	unsigned short mnc[3];
+	unsigned char tac[2];
+	unsigned char cid[4];
+	unsigned short phys_cid;
```
Are we in control of this code?

If so, can we use u16_t here?

evenl replied:
yes, but I can not use the zephyr defines as the supl library is OS agnostic. I can change these to stdint.h types.

SebastianBoe replied:
It's OS agnostic?

But we are using char2hex for instance, why is that OK, but not u32_t ?

evenl replied:
char2hex is only used in the OS integration part and is not a part of the library. supl_session.h is a part of the library. To make it cleaner I have removed the copy in nrf and it will only exist in the downloaded zip file.

carlescufi commented:
> Finished the first pass on review, I hope it's now clear where to put the library with feedback from @jhn-nordic and @carlescufi.

The only issue with placing the downloaded library in `ext/` in this repository is that users might be tempted to check it in when they run `git status`. An option is to commit a `README.txt` in `ext/lib/bin/supl` and then add `ext/lib/bin/supl` to the `.gitignore`. Thoughts?

rlubos commented:
> An option is to commit a README.txt in ext/lib/bin/supl and then add ext/lib/bin/supl to the .gitignore. Thoughts?

I agree, adding this directory to `.gitignore` should be a good precaution for committing the library accidentally.

lemrey commented on lib/bin/supllib/include/supl_session.h:
```diff
@@ -0,0 +1,317 @@
+	struct lte_params lte_params;
+
+	int device_id_choice;
+	unsigned char device_id[LIBSUPL_ID_LEN];
```
Is this string `NULL` terminated? If not, consider using `u8_t` to show that it's a byte stream instead.

evenl replied:
It's not 0 terminated, but I can not use u8_t because it's a OS agnostic library. I can change it to uint8_t though.

lemrey commented on samples/nrf9160/gps/src/supl_support.c:
```diff
@@ -0,0 +1,154 @@
+	va_end(args);
+
+	if (ret < 0) {
+		printk("%s: encoding error\n", __func__);
```
If this function is called directly, `__func__` will always resolve to `logger`.
If it is not called directly, but wrapped in a macro, I would not put `buffer` on the stack since it affects all the functions that log something.

evenl replied:
It's called through the api->logger function pointer by the SUPL client. Maybe I should call the function supl_logger or something to make less likely that this is confused by the real logger.

lemrey commented on samples/nrf9160/gps/src/supl_support.c:
```diff
@@ -0,0 +1,154 @@
+
+#define SUPL_SERVER "supl.google.com"
+
+static int            supl_fd;
```
whitespace :)

evenl replied:
ok

evenl replied:
fixed

lemrey commented:
Regarding the location, I think that if the library comes with a different license than the rest of the repo it should be put under `/ext`. If we followed the same pattern of `liblwm2m_carrier`, we could have `/ext/lib/bin`.

evenl commented on samples/nrf9160/gps/src/supl_support.c:
```diff
@@ -0,0 +1,154 @@
+		       "it will be cut short\n", __func__);
+	}
+
+	printk("%s", log_strdup(buffer));
```
fixed

evenl commented on samples/nrf9160/gps/src/supl_support.c:
```diff
@@ -0,0 +1,154 @@
+#include <zephyr.h>
```
fixed

evenl commented on samples/nrf9160/gps/src/main.c:
```diff
@@ -8,9 +8,26 @@
 #include <nrf_socket.h>
```
fixed

evenl commented on samples/nrf9160/gps/src/main.c:
```diff
@@ -252,17 +307,62 @@ int process_gps_data(nrf_gnss_data_frame_t *gps_data)
+int inject_agps_type(void *agps,
+			    size_t agps_size,
+			    nrf_gnss_agps_data_type_t type,
+			    void *user_data)
```
fixed

evenl commented on samples/nrf9160/gps/src/main.c:
```diff
@@ -244,6 +280,25 @@ int process_gps_data(nrf_gnss_data_frame_t *gps_data)
+			printk("New AGPS data requested, contacting SUPL server, flags %d\n",
+			       gps_data->agps.data_flags);
+			activate_lte(true);
+			printf("Established LTE link\n");
```
fixed

evenl commented on samples/nrf9160/gps/prj.conf:
```diff
@@ -5,14 +5,38 @@
+CONFIG_NET_TCP=n
+
+# Main thread
+CONFIG_MAIN_THREAD_PRIORITY=0
```
Removed

evenl commented on samples/nrf9160/gps/prj.conf:
```diff
@@ -5,14 +5,38 @@
+CONFIG_NET_IPV4=n
+CONFIG_NET_IPV6=n
+CONFIG_NET_UDP=n
+CONFIG_NET_TCP=n
```
yes

evenl commented on samples/nrf9160/gps/prj.conf:
```diff
@@ -5,14 +5,38 @@
+CONFIG_MAIN_THREAD_PRIORITY=0
+
+CONFIG_HEAP_MEM_POOL_SIZE=4096
+CONFIG_MAIN_STACK_SIZE=8192
```
It's probably not needed, some of the GPS data frames are large, and they are stored on the stack emediately, but they are not large. I adjusted the values.

evenl commented on lib/bin/supllib/include/supl_session.h:
```diff
@@ -0,0 +1,317 @@
+ *          <0 error
+ */
+typedef ssize_t (*supl_read_t)(
+	void *p_buff,
```
fixed

evenl commented on lib/bin/supllib/include/supl_os_client.h:
```diff
@@ -0,0 +1,25 @@
+#include <supl_session.h>
+
+/**
+ * @brief Run this function to start a SUPL session
```
Ok

evenl commented on lib/bin/supllib/include/supl_os_client.h:
```diff
@@ -0,0 +1,25 @@
+int nrf_supl_session(nrf_gnss_agps_data_frame_t *agps_request);
+
+/**
+ * @brief This function will setup the API and the buffers required by
```
ok

evenl commented on lib/bin/supllib/include/supl_os_client.h:
```diff
@@ -0,0 +1,25 @@
+#include <nrf_socket.h>
```
ok

evenl commented on lib/bin/supllib/os/lte_params.c:
```diff
@@ -0,0 +1,284 @@
+ *
+ * @return 0 when MNC filled successfully, -1 otherwise
+ */
+static int parse_lte_mnc(struct lte_params *p_lte,
```
fixed

evenl commented on lib/bin/supllib/os/supl_os_client.c:
```diff
@@ -0,0 +1,132 @@
+int nrf_supl_init(const struct supl_api *api)
+{
+	if (supl_init(api, &bufs) != 0) {
+		LOG_ERR("Failed to initialize libsupl");
```
It will always return -1 on error, it fails if the read, write and handler function pointers are NULL, if any of the buffers are NULL or if the length of any of the buffers are to small (less than LIBSUPL_PRI_BUF_SIZE).

evenl commented on lib/bin/supllib/os/supl_os_client.c:
```diff
@@ -0,0 +1,132 @@
+	return 0;
+}
+
+int nrf_supl_session(nrf_gnss_agps_data_frame_t *agps_request)
```
Yes, I'll add const before the type and after the star

evenl commented on lib/bin/supllib/os/utils.c:
```diff
@@ -0,0 +1,66 @@
+		return -1;
+	}
+
+	memset(out, 0, out_len);
```
I or the values into the out buffer, so I must know that all elements in the array is 0 before I start.

SebastianBoe commented on samples/nrf9160/gps/CMakeLists.txt:
```diff
@@ -10,4 +10,9 @@ include($ENV{ZEPHYR_BASE}/../nrf/cmake/boilerplate.cmake)
+			   src/supl_support.c)
+else()
 target_sources(app PRIVATE src/main.c)
+endif()
```
```
zephyr_library_sources(src/main.c)

zephyr_library_sources_ifdef(
CONFIG_SUPL_CLIENT_LIB
src/supl_support.c
)

```

evenl replied:
ok

rlubos commented on lib/bin/supllib/os/supl_os_client.c:
```diff
@@ -0,0 +1,138 @@
+	while (1) {
+		/* find column end */
+		const char *col_end = strchr(idx, ',');
+
```
Excess empty line.

evenl replied:
ok

rlubos commented on lib/bin/supllib/readme.rst:
```diff
@@ -0,0 +1,25 @@
+
+Installing
+**********
+Unzip the downloaded zip file to nrf/ext/bin/supl/ while keeping the folder structure
```
The CMakeLists.txt states it's `ext/supl`, not `ext/bin/supl`.

evenl replied:
ok

rlubos commented on ext/.gitignore:
```diff
@@ -0,0 +1 @@
+/supl/
```
Not sure if we should introduce a new local `.gitignore`, or use the one from the root directory, @SebastianBoe?

SebastianBoe replied:
I'd prefer to use the root directory, that makes it easier to get a global overview of ignored directories.

evenl replied:
Should have checked, wasn't aware of this file. I was also thinking that I should maybe also include libsupl.zip which is the name of the file you download.

evenl commented on lib/bin/supllib/os/lte_params.c:
```diff
@@ -0,0 +1,284 @@
+#include "lte_params.h"
+#include "utils.h"
+
+LOG_MODULE_DECLARE(supl_client, LOG_LEVEL_DBG);
```
This file only contained one line of logging so I removed logging from this file. The caller is now handling the error code.

evenl commented on lib/bin/supllib/os/supl_os_client.c:
```diff
@@ -0,0 +1,138 @@
+static struct supl_session_ctx supl_client_ctx;
+
+static struct supl_buffers bufs = {
+		.pri_buf = pri_buf,
```
ok

evenl commented:
I think I have addressed comments now and I have started on the sample documentation (new commit). Not yet done, but any comments for what information it should contain is appreciated.

lemrey commented:
Let's move the glue to `lib/supl` instead, we have a binaries under `lib/bin` which is not the case here.

ru-fu commented:
> 
> 
> Let's move the glue to `lib/supl` instead, we have a binaries under `lib/bin` which is not the case here.

This will make the documentation not show up. The doc files are usually next to the header files in the include folder. The only exception is for binary libraries that have only header files, and those are in the lib/bin folder. 
So if we move the files, the header files also need to go in the correct location (namely under include).

evenl commented:
The important API file is moved to the include folder.

ru-fu commented on include/supl_os_client.rst:
```diff
@@ -0,0 +1,35 @@
+.. _supl_client:
+
+SUPL client
```
How should this be called? We call it supllib in the bin folder and supl_os_client in the include folder ... Should this be "SUPL OS client" instead?

evenl replied:
I think SUPL client is ok. SUPLlib is the specific library component that provides the SUPL Client functionality, supl_os_client refer to the OS integration code for using this client with NCS. But all of this is parts that provides SUPL Client functionality to NCS.

ru-fu replied:
And it won't be confusing for users that the header file has a different name than the library folder?

lemrey replied:
I agree with Ruth, it would be best that the two things had the same name.

ru-fu commented on include/supl_os_client.rst:
```diff
@@ -0,0 +1,35 @@
+The Secure User-Plane Location (SUPL) client module integrates the external SUPL client library into the |NCS|.
+This library can be used to receive A-GPS data over the SUPL protocol.
+
+The SUPL client library is released under a different license than the |NCS|.
```
@bjorn-tore-taraldsen Is it a specific license that we can mention here (by name)?

bjorn-tore-taraldsen replied:
This is a proprietary licence that we made to fulfill the licence agreement we had with OSS Nokalva. So it will be hard to identify be a name. Could we add something like this to highlight that this is a different license?
"Therefore, you must download and install it separately to be able to use this module. 
The license will be made available for you to read before you download the library. Please read through it before you accept and start downloading it."

ru-fu commented on include/supl_os_client.rst:
```diff
@@ -0,0 +1,35 @@
+Downloading and installing
+**************************
+
+You can download the SUPL client library from Nordic Semiconductor's website.
```
@bjorn-tore-taraldsen Where can it be downloaded?

bjorn-tore-taraldsen replied:
Still work in progress.
Will most likely be here https://www.nordicsemi.com/Software-and-tools/Development-Kits/nRF9160-DK/Download#infotabs 
Together with the other downloads on the bottom of the page.
Suggested header for this download section: "nRF9160 SiP SUPL Client library"

ru-fu commented on include/supl_os_client.rst:
```diff
@@ -0,0 +1,35 @@
+*****************
+
+| Header file: :file:`include/supl_os_client.h`
+| Source files: :file:`lib/suplib/os/`
```
That should actually be supllib.

evenl replied:
ok

ru-fu commented on include/supl_os_client.rst:
```diff
@@ -0,0 +1,38 @@
+
+The SUPL client library is released under a different license than the |NCS|.
+Therefore, you must download and install it separately to be able to use this module.
+The license will be made available for you to read before you download the library.
```
```suggestion
Before you download the library, read through the license on the webpage.
```

ru-fu commented on include/supl_os_client.rst:
```diff
@@ -0,0 +1,38 @@
+The SUPL client library is released under a different license than the |NCS|.
+Therefore, you must download and install it separately to be able to use this module.
+The license will be made available for you to read before you download the library.
+Please read through it before you accept and start downloading it.
```
```suggestion
You must accept the license before you can download the library.
```

ru-fu commented on include/supl_os_client.rst:
```diff
@@ -0,0 +1,38 @@
+**************************
+
+You can download the SUPL client library from Nordic Semiconductor's website.
+`download https://www.nordicsemi.com/Software-and-tools/Development-Kits/nRF9160-DK/Download#infotabs`_
```
```suggestion

```

ru-fu commented on include/supl_os_client.rst:
```diff
@@ -0,0 +1,38 @@
+Downloadine and installing
+**************************
+
+You can download the SUPL client library from Nordic Semiconductor's website.
```
```suggestion
You can download the SUPL client library from the `nRF9160 DK product page`_.
```

ru-fu commented on include/supl_os_client.rst:
```diff
@@ -0,0 +1,38 @@
+You can download the SUPL client library from Nordic Semiconductor's website.
+`download https://www.nordicsemi.com/Software-and-tools/Development-Kits/nRF9160-DK/Download#infotabs`_
+
+Download the zip file and extract it into the :file:`nrf/ext/supl/` folder.
```
```suggestion
Download the nRF9160 SiP SUPL Client library zip file and extract it into the :file:`nrf/ext/supl/` folder.
```

ru-fu commented on include/supl_os_client.rst:
```diff
@@ -0,0 +1,37 @@
+Before you download the library, read through the license on the webpage.
+You must accept the license before you can download the library.
+
+Downloadine and installing
```
```suggestion
Downloading and installing
```

Oops.

ru-fu commented on include/supl_os_client.rst:
```diff
@@ -0,0 +1,37 @@
+*****************
+
+| Header file: :file:`include/supl_os_client.h`
+| Source files: :file:`lib/suplib/`
```
```suggestion
| Source files: :file:`lib/supllib/`
```

carlescufi commented on .gitignore:
```diff
@@ -26,3 +26,6 @@ doc/nrf/_doxygen/
 doc/*_warnings.txt
 doc/*_doc.log
+
+ext/supl
```
let's clean this up a bit and use `ext/lib/supl` please

ru-fu replied:
That must then be changed in the documentation as well.

evenl replied:
Ok,fixed

carlescufi commented on .gitignore:
```diff
@@ -26,3 +26,6 @@ doc/nrf/_doxygen/
 doc/*_warnings.txt
 doc/*_doc.log
+
+ext/iib/supl
```
s/iib/lib

carlescufi commented on CODEOWNERS:
```diff
@@ -66,6 +66,7 @@ Kconfig*                                  @SebastianBoe
 /lib/fatal_error/                         @ioannisg
 /lib/sms/                                 @metc
 /lib/st25r3911b/                          @grochu
+/lib/supllib/                             @rlubos @evenl
```
why not simply `/lib/supl/`?

ru-fu commented on include/supl_os_client.rst:
```diff
@@ -0,0 +1,37 @@
+Before you download the library, read through the license on the webpage.
+You must accept the license before you can download the library.
+
+Downloadine and installing
```
Why is this downloadine again?

carlescufi replied:
I assume it's a typo and should be "Downloading"
